[#1466] Fix autocomplete for aliased subcommands

rsenden opened the pull request:
This PR fixes #1466

remkop commented:
This is great, thank you!
Can you add a test for this, or modify one of the existing tests?

rsenden commented:
I just added aliases to the existing AutoComplete tests, and updated the expected outputs. The only thing that I'm not sure about is why in AutoCompleteTest#testComplete() I need to repeat each alias three times. Any idea?

remkop commented:
> The only thing that I'm not sure about is why in AutoCompleteTest#testComplete() I need to repeat each alias three times. Any idea?

I will have to take a look in the debugger to see what is going on. It may take some time before I can get to this.

remkop commented:
I finally had some time to take a look at this.
It appears that the `AutoComplete#filterAndTrimMatchingPrefix` method does not filter out duplicates, even though it should. This can be fixed by modifying the first line of this method to use a `Set` instead of a `List`:

```java
    private static void filterAndTrimMatchingPrefix(String prefix, List<CharSequence> candidates) {
        Set<CharSequence> replace = new HashSet<CharSequence>();
        ...
```

The expectations in `testComplete` can then be changed to also have only one occurrence of each of the subcommand aliases, which makes much more sense.

Can you change the PR accordingly?

remkop commented on src/test/resources/picocompletion-demo-help_completion.bash:
```diff
@@ -4,7 +4,7 @@
 #
 # Bash completion support for the `picocompletion-demo-help` command,
-# generated by [picocli](http://picocli.info/) version %1$s.
+# generated by [picocli](http://picocli.info/) version 4.6.3-SNAPSHOT.
```
Please leave the `%1$s` variable, it will be replaced by the latest version in the test.

remkop commented on src/test/resources/picocompletion-demo-help_completion.bash:
```diff
@@ -275,7 +396,7 @@ function _picocli_picocompletion-demo-help_sub2_subsub2() {
   local flag_opts=""
   local arg_opts="-u --timeUnit -t --timeout"
-  local timeUnit_option_args="%2$s" # --timeUnit values
+  local timeUnit_option_args="NANOSECONDS MICROSECONDS MILLISECONDS SECONDS MINUTES HOURS DAYS" # --timeUnit values
```
Please leave the `%2$s` variable, it will be replaced by the supported TimeUnit constants for the version of Java that is running the test.

remkop commented on src/test/resources/picocompletion-demo_completion.bash:
```diff
@@ -4,7 +4,7 @@
 #
 # Bash completion support for the `picocompletion-demo` command,
-# generated by [picocli](http://picocli.info/) version %1$s.
+# generated by [picocli](http://picocli.info/) version 4.6.3-SNAPSHOT.
```
Please leave the `%1$s` variable, it will be replaced by the latest version in the test.

remkop commented on src/test/resources/picocompletion-demo_completion.bash:
```diff
@@ -256,7 +377,7 @@ function _picocli_picocompletion-demo_sub2_subsub2() {
   local flag_opts=""
   local arg_opts="-u --timeUnit -t --timeout"
-  local timeUnit_option_args="%2$s" # --timeUnit values
+  local timeUnit_option_args="NANOSECONDS MICROSECONDS MILLISECONDS SECONDS MINUTES HOURS DAYS" # --timeUnit values
```
Please leave the `%2$s` variable, it will be replaced by the supported TimeUnit constants for the version of Java that is running the test.

remkop reviewed (changes requested):
I also found some minor things with variables used in the test result templates that should be preserved. Can you take a look?

rsenden commented:
PR has been updated with all requested changes

remkop commented:
Merged. 
Thank you for the contribution!

I will update the release notes when I get to my PC.
